// purescript-psa/output/Global/index.js
// Generated by purs version 0.13.8

// | This module defines types for some global Javascript functions
// | and values.
"use strict";
var $foreign = require("./foreign.js");
var Data_Function = require("../Data.Function/index.js");
var Data_Maybe = require("../Data.Maybe/index.js");

// | Formats Number as String in fixed-point or exponential notation rounded
// | to specified number of significant digits. May return `Nothing` when
// | precision is less than 1 or greater than 21 depending on the
// | implementation. See ECMA-262 for more information.
var toPrecision = function (digits) {
    return function (n) {
        return $foreign["_toPrecision"](Data_Function["const"](Data_Maybe.Nothing.value), Data_Maybe.Just.create, digits, n);
    };
};

// | Formats Number as a String with limited number of digits after the dot.
// | May return `Nothing` when specified number of digits is less than 0 or
// | greater than 20. See ECMA-262 for more information.
var toFixed = function (digits) {
    return function (n) {
        return $foreign["_toFixed"](Data_Function["const"](Data_Maybe.Nothing.value), Data_Maybe.Just.create, digits, n);
    };
};

// | Formats Number as String in exponential notation limiting number of digits
// | after the decimal dot. May return `Nothing` when specified number of
// | digits is less than 0 or greater than 20 depending on the implementation.
// | See ECMA-262 for more information.
var toExponential = function (digits) {
    return function (n) {
        return $foreign["_toExponential"](Data_Function["const"](Data_Maybe.Nothing.value), Data_Maybe.Just.create, digits, n);
    };
};

// | URI component encoding. Returns `Nothing` when given a value with
// | unencodeable characters.
var $$encodeURIComponent = function (s) {
    return $foreign["_encodeURIComponent"](Data_Function["const"](Data_Maybe.Nothing.value), Data_Maybe.Just.create, s);
};

// | URI encoding. Returns `Nothing` when given a value with unencodeable
// | characters.
var $$encodeURI = function (s) {
    return $foreign["_encodeURI"](Data_Function["const"](Data_Maybe.Nothing.value), Data_Maybe.Just.create, s);
};

// | URI component decoding. Returns `Nothing` when given a value with
// | undecodeable escape sequences.
var $$decodeURIComponent = function (s) {
    return $foreign["_decodeURIComponent"](Data_Function["const"](Data_Maybe.Nothing.value), Data_Maybe.Just.create, s);
};

// | URI decoding. Returns `Nothing` when given a value with undecodeable
// | escape sequences.
var $$decodeURI = function (s) {
    return $foreign["_decodeURI"](Data_Function["const"](Data_Maybe.Nothing.value), Data_Maybe.Just.create, s);
};
module.exports = {
    toFixed: toFixed,
    toExponential: toExponential,
    toPrecision: toPrecision,
    "decodeURI": $$decodeURI,
    "encodeURI": $$encodeURI,
    "decodeURIComponent": $$decodeURIComponent,
    "encodeURIComponent": $$encodeURIComponent,
    nan: $foreign.nan,
    "isNaN": $foreign["isNaN"],
    infinity: $foreign.infinity,
    "isFinite": $foreign["isFinite"],
    readInt: $foreign.readInt,
    readFloat: $foreign.readFloat
};
